### ISSUE TITLE
Open Question: Good ecosystem selections for parsing toolchain?

### ISSUE THREAD

**Issue by Swiddis** [Swiddis/ferrous-qat#19]
There are [a lot of crates for parser tooling](https://lib.rs/parsing). What are good ecosystem selections? I'm currently looking at switching to [Logos](https://lib.rs/crates/logos) for the lexer and [Pest](https://lib.rs/crates/pest) for the parser, but there's also options like [nom](https://lib.rs/crates/nom) that claims to handle both lexing and parsing.

I'll probably experiment a little with the top crates at some point to see how the results look.

**Comment by Swiddis** [comment]
Example Pest spec for non-recursive expressions:

```pest
// Constant sets
letter    = { 'a'..'z' }
any       = { "." }
consonant = { "#" }
vowel     = { "@" }

// Sets
set_range = { letter ~ "-" ~ letter }
set       = { "[" ~ (set_range | letter)+ ~ "]" }
negset    = { "[!" ~ (set_range | letter)+ ~ "]" }

// Non-recursive expression components
fixed_len = { (any | consonant | vowel | set | negset | letter)+ }
wildcard  = { "*" }
word      = { ">" }
rev_word  = { "<" }

basic_component = { fixed_len | wildcard | word | rev_word }

// If no length restriction, match length must eq anagram length.
// Otherwise, anagram is a superset of matched word
anagram = { "/" ~ fixed_len }
// Subset of match instead of superset
must_contain = { "/" ~ fixed_len ~ "*" }
// Exactly one non-any is wrong
hamming = { "`" ~ fixed_len }
// At most one non-any is wrong
hamming_opt = { "?`" ~ fixed_len }
// Length restrictions
digit     = { '0'..'9' }
len_min   = { digit+ }
len_max   = { digit+ }
len_range = { ((len_min? ~ "-" ~ len_max?) | digit+) ~ ":" }

full_component = { len_range? ~ ((basic_component+ ~ (hamming | anagram)?) | hamming | anagram) }

expr = { full_component }
```

**Comment by Swiddis** [comment]
Meanwhile an experimental lexer branch is open, [`logos-lex`](https://github.com/Swiddis/ferrous-qat/tree/logos-lex).

**Comment by Swiddis** [comment]
Pest seems like a good choice, tracking work in #18

**Comment by Swiddis** [comment]
Marking as closed by #20
